lib/parse.js

```diff
@@ -2506,56 +2546,7 @@ function parse($TEXT, options) {
         var val = S.token.value;
 
         if (is("operator") && ASSIGNMENT(val)) {
-            if (is_assignable(left)) {
-
-                var walk = function(node) {
-                    var newNode;
-                    if (node instanceof AST_Object) {
-                        newNode = new AST_Destructuring({
-                            start: node.start,
-                            names: node.properties.map(walk),
-                            is_array: false,
-                            end: node.end
-                        });
-                        node = newNode;
-                    } else if (node instanceof AST_Array) {
-                        var names = [];
-
-                        for (var i = 0; i < node.elements.length; i++) {
-                            // Only allow expansion as last element
-                            if (node.elements[i] instanceof AST_Expansion) {
-                                if (i + 1 !== node.elements.length) {
-                                    token_error(node.elements[i].start, "Spread must the be last element in destructuring array");
-                                }
-                                node.elements[i].expression = walk(node.elements[i].expression);
-                            }
-
-                            names.push(walk(node.elements[i]));
-                        }
-
-                        newNode = new AST_Destructuring({
-                            start: node.start,
-                            names: names,
-                            is_array: true,
-                            end: node.end
-                        });
-                        node = newNode;
-                    } else if (node instanceof AST_ObjectProperty) {
-                        node.value = walk(node.value);
-                    } else if (node instanceof AST_Assign) {
-                        node = new AST_DefaultAssign({
-                            start: node.start,
-                            left: node.left,
-                            operator: "=",
-                            right: node.right,
-                            end: node.end
-                        });
-                    }
-
-                    return node;
-                }
-                left = walk(left);
-
+            if (is_assignable(left) || (left = to_destructuring(left)) instanceof AST_Destructuring) {
```

Comment:
https://github.com/mishoo/UglifyJS2/pull/1627#issuecomment-288637954

`is_assignable()` is also used for checking the expression on `++`/`--`, so only allows `AST_Destructuring` here instead of putting this conditional into `is_assignable()`.